pkg/services/ngalert/schedule/alert_rule_test.go

```diff
@@ -762,8 +763,95 @@ func TestRuleRoutine(t *testing.T) {
+		sender := NewSyncAlertsSenderMock()
+		sender.EXPECT().Send(mock.Anything, rule.GetKey(), mock.Anything).Return()
+
+		sch, ruleStore, _, _ := createSchedule(evalAppliedChan, sender)
+		sch.stateManager.ResolvedRetention = 4 * time.Second
+		sch.stateManager.ResendDelay = 2 * time.Second
+		sch.stateManager.Put([]*state.State{
+			stateForRule(t, rule, sch.clock.Now(), eval.Alerting), // Add existing Alerting state so evals will resolve.
+		})
+
+		ruleStore.PutRule(context.Background(), rule)
+		factory := ruleFactoryFromScheduler(sch)
+		ctx, cancel := context.WithCancel(context.Background())
+		t.Cleanup(cancel)
+		ruleInfo := factory.new(ctx, rule)
+
+		go func() {
+			_ = ruleInfo.Run(rule.GetKey())
+		}()
+
+		// Evaluate 10 times:
+		// 1. Send resolve #1.
+		// 2. 2s resend delay.
+		// 3. Send resolve #2.
+		// 4. 2s resend delay.
+		// 5. Send resolve #3.
+		// 6. No more sends, 4s retention period is over.
+		expectedResolves := map[time.Time]struct{}{
+			sch.clock.Now().Add(1 * time.Second): {},
+			sch.clock.Now().Add(3 * time.Second): {},
+			sch.clock.Now().Add(5 * time.Second): {},
+		}
+		calls := 0
+		for i := 1; i < 10; i++ {
+			ts := sch.clock.Now().Add(time.Duration(int64(i)*rule.IntervalSeconds) * time.Second)
+			ruleInfo.Eval(&Evaluation{
+				scheduledAt: ts,
+				rule:        rule,
+			})
+			waitForTimeChannel(t, evalAppliedChan)
+
+			if _, ok := expectedResolves[ts]; ok {
+				calls++
+				prevCallAlerts, ok := sender.Calls()[calls-1].Arguments[2].(definitions.PostableAlerts)
+				assert.Truef(t, ok, fmt.Sprintf("expected argument of function was supposed to be 'definitions.PostableAlerts' but got %T", sender.Calls()[calls-1].Arguments[2]))
+				assert.Len(t, prevCallAlerts.PostableAlerts, 1)
+			}
+			sender.AssertNumberOfCalls(t, "Send", calls)
+		}
+	})
 }
 
 func ruleFactoryFromScheduler(sch *schedule) ruleFactory {
 	return newRuleFactory(sch.appURL, sch.disableGrafanaFolder, sch.maxAttempts, sch.alertsSender, sch.stateManager, sch.evaluatorFactory, &sch.schedulableAlertRules, sch.clock, sch.featureToggles, sch.metrics, sch.log, sch.tracer, sch.recordingWriter, sch.evalAppliedFunc, sch.stopAppliedFunc)
 }
+
+func stateForRule(t *testing.T, rule *models.AlertRule, ts time.Time, evalState eval.State) *state.State {
+	s := &state.State{
+		OrgID:              rule.OrgID,
+		AlertRuleUID:       rule.UID,
+		CacheID:            "",
+		State:              evalState,
+		Annotations:        make(map[string]string),
+		Labels:             make(map[string]string),
+		StartsAt:           ts,
+		EndsAt:             ts,
+		ResolvedAt:         ts,
+		LastSentAt:         ts,
+		LastEvaluationTime: ts,
+	}
+	for k, v := range rule.Labels {
+		s.Labels[k] = v
+	}
+	for k, v := range state.GetRuleExtraLabels(&logtest.Fake{}, rule, "", true) {
+		if _, ok := s.Labels[k]; !ok {
+			s.Labels[k] = v
+		}
+	}
+	il := models.InstanceLabels(s.Labels)
+	s.Labels = data.Labels(il)
+	id, err := il.StringKey()
+	require.NoError(t, err)
```

Comment:
This has just been changed in another PR. Let's update to use labels' fingerprint